### PR TITLE
Reorg vfu_create_ctx() and also create dma_controller

### PR DESCRIPTION
**Pull request by swapnili**
dma_controller_add_region() adds dma_region only if
vfu_ctx->dma is created.

vfu_setup_device_dma_cb() should only set callbacks.

Signed-off-by: Swapnil Ingle <swapnil.ingle@nutanix.com>

### CODE REVIEW

File: lib/libvfio-user.c [hunk 1]
```diff
@@ -1334,7 +1335,7 @@ vfu_create_ctx(vfu_trans_t trans, const char *path, int flags, void *pvt,
     vfu_ctx->uuid = strdup(path);
     if (vfu_ctx->uuid == NULL) {
         err = errno;
```
Comment (tmakatos) [inline lib/libvfio-user.c:1337]: `-errno` (since you're touching this code)

**Review by tmakatos — CHANGES_REQUESTED** [review]
According to the API documentation the user must explicitly ask for a DMA controlling by registering a callabck. Not registering a callback means that the device isn't going to perform DMA to the guest, so DMA map/unmap commands are simply ack'ed. Why are we changing this behavior?

**Comment by swapnili** [comment]
Yes you are right there is no need to keep the maps if user do not want them.
From the documentation of vfu_setup_device_dma_cb() it was not clear, I will add some explanation there.
I think if vfu_setup_device_dma_cb() is called we should unconditionally dma_controller_create()?

File: lib/libvfio-user.c (hunk 1 above)
Comment (swapnili) [reply lib/libvfio-user.c:1337]: Ack

**Comment by tmakatos** [comment]
> I think if vfu_setup_device_dma_cb() is called we should unconditionally dma_controller_create()?

Exactly.

**Review by tmakatos — APPROVED** [review]

**Pull request merged**